### ISSUE TITLE
[TS] Introduce type testing setup

### ISSUE THREAD

**Issue by kacper-mikolajczak** [Expensify/App#46639]
### **Problem**

Originating from https://github.com/Expensify/App/pull/46280, there is a realisation that some types are complex and potentially may not work properly within all edge cases.

### **Solution**

We can guard against such regressions by using one of the tools that offers type testing functionality:
- [tsd](https://github.com/tsdjs/tsd) - library consist of API for unit testing TS types
- [vitest type testing module](https://vitest.dev/guide/testing-types.html) - enables writing a test suites with Vitest API

**Comment by kacper-mikolajczak** [comment]
Hey hey @roryabraham 👋 

When making research for type testing setup, I thought if we could simplify such testing a bit.

1. Adding yet another 3rd party with scripts and all the setup is quite meh.
2. We already have a TS checker on pipeline running, which ensures all types are defined correctly
3. Can we utilise above functionality for the purpose of testing by writing tests as types?

We can create a framework out of couple TS testing helpers like ones below:

```ts
/**
 * Test suite type. It defines a framework for testing. Simple yet everything we need to verify tests pass.
 * Key of the record is a test description and the value becomes actual test that needs to evaluate to true
 */
type TestSuite = Record<string, true>;

/**
 * One of the assertion helpers. Analogous to jest helpers. We can create more like `NotEquals` etc. 
 */
type Equals<T, U> = (<G>() => G extends T ? 1 : 2) extends <G>() => G extends U ? 1 : 2 ? true : false;
```

Equipped with those, we could potentially create tests out of it. As an example, let's take `NonPartial` type definition and try to write tests for it:

```ts
/**
 * Utility type to make all properties of an object non-optional with `undefined` as a possible value. It fixes the issue with `Required` utility type, which makes all properties required.
 * See https://github.com/microsoft/TypeScript/issues/31025 for more details.
 */
type NonPartial<T, RT extends T = Required<T>> = {[K in keyof RT]: K extends keyof T ? (undefined extends T[K] ? T[K] | undefined : T[K]) : never};
```

The testing would look like so:

```ts
// NonPartial.test.d.ts
import type { Equals, TestSuite } from './TypeTestingHelpers';
import type NonPartial from './NonPartial';

/**
 * Test suite
 */
interface NonPartialTestSuite extends TestSuite {
    'should replace optional with undefined': Equals<NonPartial<{optional?: number}>, {optional: number | undefined}>; // OK
    'should keep undefined as is': Equals<NonPartial<{withUndef: number | undefined}>, {withUndef: number | undefined}>; // OK
    'should fail on CI': Equals<NonPartial<{withUndef: number | undefined}>, {withUndef: number}>; // FAILS PIPELINE CHECK
}
```

For types there are no complex behaviour (like integration tests) so mechanism that verifies their structural correctness should be enough. Please let me know what you think!

**Comment by kacper-mikolajczak** [comment]
As an alternative, I was thinking about `Test` type as well which could work like so:

```ts
// TypeTestingHelpers.d.ts
type Test = {
    it: string;
    input: unknown;
    expected: unknown;
    test: true;
};
...

// NonPartial.test.d.ts
import type { Equals, TestSuite } from './TypeTestingHelpers';
import type NonPartial from './NonPartial';

interface T1 extends Test {
    it: 'should replace optional with undefined';
    input: {
        optional?: number;
    };
    expected: {
        optional: number | undefined;
    };
    test: Equals<NonPartial<T1['input']>, T1['expected']>;
}

interface T2 extends Test {
    it: 'should keep undefined as is';
    input: {
        withUndef: number | undefined;
    };
    expected: {
        withUndef: number | undefined;
    };
    test: Equals<NonPartial<T2['input']>, T2['expected']>;
}

interface T3 extends Test {
    it: 'should fail on CI';
    input: {
        withUndef: number | undefined;
    };
    expected: {
        withUndef: number;
    };
    test: Equals<NonPartial<T3['input']>, T3['expected']>;
}
```

Pros:
- better test encapsulation
- `input` and `output` are defined separately, which makes it easier to understand and modify the test if needed

Cons:
- verbose (needs to write `extends Test` clause and make a name for the type)

**Comment by roryabraham** [comment]
I like both of those ideas. Personally I prefer the 2nd option because, even though it's a bit more verbose, it's a bit clearer/more readable imho. Either one works 👍🏼

**Comment by roryabraham** [comment]
(also, we don't need to go crazy adding tests for all our types, just some "pilot" tests would be good. That way they can be a useful developer tool if you're ever adding or modifying complex types in the future)

**Comment by adhorodyski** [comment]
Hey! (Coming from the original thread)

We talked about this in the team and I wanted to post a summary of what we came up with:
- _testing types_ is useful in the library land, not so much the application code where we mostly utilise types
- `tsc` (Typescript's compiler) takes care of all the usages for us. It will block the wrong usage, flag a new edge case in the IDE (if we have any) and yell at us on the CI if necessary. It in fact is a real-time and CI testing layer for types and how we use them.
- such an initiative goes over creating a new internal tool/small library/framework that we'd then have to eventually maintain (design, create, document, bugfix etc), and this comes at a huge cost that is spreaded all across the timeline.

TLDR this isn't a question of **testing types VS not testing the types**, but `do we want to additionally test types already checked by tsc?`.

For me it's not really worth it, I'd not vote for `not testing the types`, but for **not having additional type testing layer on top of the `tsc`** that already checks it for us.
